internal/config/dynamic_config_watcher.go

```diff
@@ -0,0 +1,92 @@
+package config
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+type  DynamicConfigWatcher struct {
+	ticker *time.Ticker
+	exists bool
+	hash string
+}
+
+func NewDynamicConfigWatcher(canaryConfig *CanaryConfig, applyFunc func(config *DynamicCanaryConfig), defaultFunc func() (*DynamicCanaryConfig)) (*DynamicConfigWatcher, error) {
+	dynamicConfigWatcher := &DynamicConfigWatcher{}
+
+	if canaryConfig.DynamicConfigFile != "" && canaryConfig.DynamicConfigWatcherInterval > 0 {
+		glog.Infof("Starting dynamic config watcher for file %s with period %d ms", canaryConfig.DynamicConfigFile, canaryConfig.DynamicConfigWatcherInterval)
+
+		// Apply any existing config from the file-system
+		target, hsh, err := readAndHash(canaryConfig.DynamicConfigFile)
+		if err == nil && target != nil {
+			dynamicConfigWatcher.hash = hsh
+			dynamicConfigWatcher.exists = true
+			applyFunc(target)
+		}
+
+		dynamicConfigWatcher.ticker = time.NewTicker(canaryConfig.DynamicConfigWatcherInterval * time.Millisecond)
+		quit := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case <- dynamicConfigWatcher.ticker.C:
+					if _, err := os.Stat(canaryConfig.DynamicConfigFile); err == nil {
+						dynamicConfigWatcher.exists = true
+						target, hsh, err := readAndHash(canaryConfig.DynamicConfigFile)
+						if err != nil || target == nil {
+							glog.Warningf("failed to read and hash %s : %v (ignored)", canaryConfig.DynamicConfigFile, err)
+							continue
+						}
+						if hsh == dynamicConfigWatcher.hash {
+							continue
+						}
+						dynamicConfigWatcher.hash = hsh
+						applyFunc(target)
+					} else if errors.Is(err, os.ErrNotExist) && dynamicConfigWatcher.exists {
+						dynamicConfigWatcher.exists = false
+						applyFunc(defaultFunc())
+					}
+				case <- quit:
+					dynamicConfigWatcher.ticker.Stop()
+					return
+				}
+			}
+		}()
+	}
+
+	return dynamicConfigWatcher, nil
+}
+
+func (c DynamicConfigWatcher) Close()  {
+	if c.ticker != nil {
+		c.ticker.Stop()
+		c.ticker = nil
+	}
+}
+
+func readAndHash(filename string) (target *DynamicCanaryConfig, h string, err error) {
+	byteValue, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return
```

Comment:
are we sure that this return statement is returning the `err` you got from the `ReadFile`?
The previous line uses `:=` (so declare and assign) so I would expect that the `err` you are checking with `nil` is the local variable but then returning the `err` named return value which is `nil`.
Or this local `err` variable is hiding the return named one and it's returned instead of it? Not sure about it ...

Reply:
Yes.  Golang's shorthand declaration declares only variables that don't already exist, so err on the line 75 is the same as err on line 74. 

I can remove the use of the naked return if you prefer.  Originally `read` and `hash` were two separate short functions.  In the longer function maybe the naked return just obfuscates.

Reply:
> Yes. Golang's shorthand declaration declares only variables that don't already exist, so err on the line 75 is the same as err on line 74.

Right!

> I can remove the use of the naked return if you prefer.

I am fine tbh.